test/xds/xds_client_ack_nack_test.go

```diff
@@ -165,9 +147,31 @@ func (s) TestClientResourceVersionAfterStreamRestart(t *testing.T) {
 	lis.Restart()
 
 	// Wait for all the previously sent resources to be re-requested.
-	ackVersionsAfterRestart, err := readResourceVersions(ctx, resourcesRequestedAfterStreamClose)
-	if err != nil {
-		t.Fatal(err)
+	ackVersionsAfterRestart := make(map[string]string)
+AcksAfterRestart:
+	for {
+		select {
+		case tuple := <-requestsCh:
+			if tuple.id == idBeforeRestart {
+				// Ignore any stray requests from the old stream.
+				continue
+			}
+			if tuple.id != idAfterRestart {
+				t.Fatalf("Received request with stream ID %d, expecting %d", tuple.id, idAfterRestart)
+			}
+			// After stream closure, capture the first request for every resource.
+			// This should not be set to an empty version string, but instead should
+			// be set to the version last ACKed before stream closure.
+			req := tuple.req
+			if len(req.GetResourceNames()) != 0 {
+				ackVersionsAfterRestart[req.GetTypeUrl()] = req.GetVersionInfo()
+			}
+			if len(ackVersionsAfterRestart) == wantResources {
+				break AcksAfterRestart
+			}
+		case <-ctx.Done():
+			t.Fatal("timeout when waiting for resources to be re-requested after stream restart")
+		}
 	}
```

Comment:
Can we do this in the loop above? I.e. store the first `wantResources` requests for each stream ID.  Could either be `[<id int>]map[string]string` or `map[<id int>]map[string]string`, or keep it as two different maps.

You could potentially go even further: get rid of the channel entirely, accumulate the maps directly in the callback instead, and close a channel / fire an event / cancel a context (can have the timeout on it) when the maps have been populated.

Reply:
Done. Moved the logic to populate the map into the callback. The test goroutine now waits for a signal from the callback to go ahead and verify the acks before and after stream restart. 

The logic to extract and compare the acks from the populated map seems a little verbose. Not sure if I can do any better though. Thanks.